### ISSUE TITLE
Python code runner efficiency is degrading heavily with time

### ISSUE THREAD

**Issue by FArekkusu** [Codewars/codewars-runner-cli#757]
Something is wrong with the code runner (or is it CW?) when it comes to katas in Python, because it keeps getting slower. A few times in the last few months I've run into issues about the tests being impossible to pass because the user was left with less time than their solution could realistically run for. [This kata](https://www.codewars.com/kata/5765870e190b1472ec0022a2/discuss) is an example of this. Around a year ago I translated it to Python: the test generation + reference solution took 3.5 seconds. Today I had to change the tests because the original solution couldn't even reach the intensive cases.

**Comment by kazk** [comment]
The translation was published on `2018-04-12T18:35:43.988Z` and finished in `6633` ms. We can't directly compare because random inputs are generated, but the same code takes almost twice as long.

I'm not sure what caused this. Python is not changed often.

Changes made to Python during this period are:

- Update Python patch versions
  - `2.7.14` -> `2.7.15`
  - `3.4.8` -> `3.4.9`
  - `3.6.4` -> `3.6.8`
  - These includes security patches and might have caused some performance issues that affected this translation.
- Add `regex` package
  - I don't think this is relevant
- Update test framework
  - created https://github.com/Codewars/python-test-framework
  - I don't think this is relevant

I don't think anything else that we control changed.

**Comment by Voileexperiments** [comment]
What about the runner environment? Did the specs change or something?

**Comment by kazk** [comment]
What do you mean by the runner environment? I haven't changed the configuration of VMs that containers are hosted in if that's what you mean.

Container images we use are derived from [the official ones](https://hub.docker.com/_/python) and the only difference we make is to add an user and install packages. I'll try some challenges that timeouts in the official images without these changes and see if it's reproducible.

**Comment by kazk** [comment]
"Smallest possible sum" takes about 14s using the official image on my computer.

If you want to try, create some folder to work in and do
```shell
# get the exact version of the test framework
wget https://raw.githubusercontent.com/Codewars/python-test-framework/5e959d8a4e86811be7ec1dfe488b106b939951b5/cw-2.py
# rename to make importing easy
mv cw-2.py cw_2.py
```
Then save solution in `solution.py` and tests in `tests.py`.
Prepend the following to `tests.py` to make it import solution and test framework.
```python
from solution import *
import cw_2 as test
```
Then do
```shell
docker run --rm -it -v "$PWD":/workspace -w /workspace python:3.4.9-jessie python tests.py
```
This takes about 14s on my computer. I also tried the previous version `3.4.8` and it's about the same. So looks like it's not Python version that's causing it.

---

I forgot that Docker version was updated from `18.06.1~ce~3-0~ubuntu` to `18.06.2~ce~3-0~ubuntu` around 6 months ago for a security patch. The issue persists with the version on my computer (`18.09.8-ce`) so anything newer might be affected. I'll check what changed.

**Comment by kazk** [comment]
I haven't figured out the exact cause, but it's not the runner. Maybe some internal changes to Docker affected Python's performance.

Some functions are extremely slow inside a container. For example, `randrange` is almost twice as slow.
```python
# randrange_perf.py extracted from tests for "Smallest possible sum"
from random import randrange
for _ in range(100): [randrange(50,500) for _ in range(randrange(30000,50000))]
```

#### Python 3.7.3

```shell
python -m cProfile -s tottime randrange_prof.py
```
```
         16686919 function calls (16686876 primitive calls) in 5.340 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
  4031891    2.111    0.000    4.524    0.000 random.py:174(randrange)
  4031891    1.681    0.000    2.413    0.000 random.py:224(_randbelow)
      100    0.771    0.008    5.295    0.053 randrange_prof.py:2(<listcomp>)
```

#### Python 3.7.3 in container

```shell
docker run --rm -it -v "$PWD":/workspace -w /workspace python:3.7.3-stretch python -m cProfile -s tottime randrange_prof.py
```
```
         16574913 function calls (16574868 primitive calls) in 10.201 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
  4005259    4.301    0.000    8.670    0.000 random.py:174(randrange)
  4005259    3.302    0.000    4.369    0.000 random.py:224(_randbelow)
      100    1.481    0.015   10.151    0.102 randrange_prof.py:2(<listcomp>)
```

**Comment by kazk** [comment]
I still don't know the cause of this yet. I couldn't find any issues opened related to performance inside container.

---

By the way, I didn't realize how fast `pypy` is. The same code using `pypy3` in a container:
```
docker run --rm -it -v "$PWD":/workspace -w /workspace pypy:3.6-slim-stretch pypy3 -m cProfile -s tottime randrange_prof.py
```
```
         20892426 function calls (20891853 primitive calls) in 1.136 seconds

   Ordered by: internal time

   ncalls  tottime  percall  cumtime  percall filename:lineno(function)
  4062257    0.281    0.000    0.566    0.000 random.py:229(_randbelow)
      100    0.250    0.003    0.990    0.010 randrange_prof.py:3(<listcomp>)
  4622386    0.245    0.000    0.245    0.000 {method 'getrandbits' of 'Random' objects}
  4062257    0.107    0.000    0.741    0.000 random.py:223(randint)
       20    0.086    0.004    0.086    0.004 {built-in function compile}
  4062257    0.069    0.000    0.635    0.000 random.py:179(randrange)
  4062259    0.040    0.000    0.040    0.000 {method 'bit_length' of 'int' objects}
```

**Comment by KamuiYato** [comment]
Hello, the problem is still here (kata : smallest sum possible) I always get timeout whatever code I write

**Comment by Steffan153** [comment]
@KamuiYato Yeah, the issue is still open, so the problem is obviously still here :wink:

**Comment by KamuiYato** [comment]
> 
> 
> @KamuiYato Yeah, the issue is still open, so the problem is obviously still here 😉

excuse me, I'm new to GitHub thank you for pointing this out !

**Comment by patrickDBL** [comment]
Hi, this problem appears on the following kata:
https://www.codewars.com/kata/ludicrous-coloured-triangles/

Even a trivial function that returns 0 times out.

**Comment by kazk** [comment]
I found a [blog post](https://medium.com/better-programming/faster-python-in-docker-d1a71a9b9917) describing how much Docker's security features affect Python's performance inside container. It matches my local experiments (almost being twice as slow).

![image](https://user-images.githubusercontent.com/639336/77383214-9649af00-6d3f-11ea-90c8-95f6c5fe72b6.png)


For us, this performance degradation happened when I updated Docker version from `18.06.1~ce~3-0~ubuntu` to `18.06.2~ce~3-0~ubuntu` for some security patches. So it seems like some security changes that was introduced there has been affecting Python's performance greatly.

Since we're always running untrusted code, we can't disable all of the security to gain speed. I'll experiment with some configurations to see if there's a way to make it faster without disabling everything.

---

Also, the post mentions that highly optimized code using `numpy` wasn't affected as much:

![image](https://user-images.githubusercontent.com/639336/77383344-ecb6ed80-6d3f-11ea-99c1-4960949672a4.png)

In our case, it seems like most of the kata suffering with this issue is generating lots of random numbers using a loop, so maybe using `numpy` to generate at once will help.

See also https://eli.thegreenplace.net/2018/slow-and-fast-methods-for-generating-random-integers-in-python/ for how `numpy` can be efficient.

**Comment by kazk** [comment]
The root cause seems to be `libseccomp` v2.4.x (https://github.com/seccomp/libseccomp/issues/153). v2.4.0 included a security fix (https://github.com/seccomp/libseccomp/issues/139) that apparently caused some serious performance regression.
We currently have v2.4.1 installed.

They have a fix https://github.com/seccomp/libseccomp/pull/180 that's supposed to be included in v2.5.0 so we can update once that's released.

**Comment by hashar** [comment]
For what it is worth, I have encountered the same python slowdown with libseccomp2 2.3.1 (Debian Stretch) or 2.3.3 (Debian Buster).
